src/extension/infer.rs

```diff
@@ -109,48 +110,71 @@ pub enum InferExtensionError {
 
 /// A graph of metavariables which we've found equality constraints for. Edges
 /// between nodes represent equality constraints.
-struct EqGraph {
-    equalities: pg::Graph<Meta, (), petgraph::Undirected>,
+struct GraphContainer<Dir: EdgeType> {
+    graph: pg::Graph<Meta, (), Dir>,
     node_map: HashMap<Meta, pg::NodeIndex>,
 }
 
-impl EqGraph {
-    /// Create a new `EqGraph`
-    fn new() -> Self {
-        EqGraph {
-            equalities: pg::Graph::new_undirected(),
-            node_map: HashMap::new(),
+macro_rules! impl_graph_container {
+    ($dir:ty) => {
+        impl GraphContainer<$dir> {
```

Comment:
I presume you can't do sthg like
```
impl<T: petgraph::EdgeType> GraphContainer<T> { ... }
```

Reply:
Hmm, I wrote this macro before I found out about the `EdgeType` trait, so I'll see if I can redo it!